### PR TITLE
Introduce rolling upgrade test

### PR DESCRIPTION
**Pull request by amoskong**
This PR introduced a new rolling upgrade, which uses new scenario with multiple kinds of c-s workload. The patchset also introduced a helper to rollback node, and support different procedure according to the version.

### CODE REVIEW

File: upgrade_test.py [hunk 1]
```diff
@@ -31,50 +31,105 @@ class UpgradeTest(FillDatabaseData):
...
     :avocado: enable
     """
+    orig_ver = None
+    new_ver = None
 
     def upgrade_node(self, node):
         repo_file = self.db_cluster.params.get('repo_file', None,  'scylla.repo.upgrade')
         new_version = self.db_cluster.params.get('new_version', None,  '')
         upgrade_node_packages = self.db_cluster.params.get('upgrade_node_packages')
+        upgrade_mode = self.db_cluster.params.get('upgrade_mode', default='smooth')
```
Comment (roydahan) [inline upgrade_test.py]: mode shouldn't be a param, it should be defined according to the base version and the target target version (can parse it from the repo target).
Something in the spirit of:
if base == opensource & target == enterprise:
   mode = reinstall (direct)
else:
   mode = smooth

File: upgrade_test.py [hunk 2]
```diff
@@ -31,50 +31,105 @@ class UpgradeTest(FillDatabaseData):
...
+            # flush all memtables to SSTables
+            node.remoter.run('sudo nodetool drain')
             node.remoter.run('sudo nodetool snapshot')
+            node.remoter.run('sudo systemctl stop scylla-server.service')
             # update *development* packages
             node.remoter.run('sudo rpm -UvhR --oldpackage /tmp/scylla/*development*', ignore_status=True)
             # and all the rest
             node.remoter.run('sudo rpm -URvh --replacefiles /tmp/scylla/* | true')
             node.remoter.run('rpm -qa scylla\*')
         elif repo_file:
```
Comment (roydahan) [inline upgrade_test.py]: no need for this if

File: upgrade_test.py [hunk 3]
```diff
@@ -31,50 +31,105 @@ class UpgradeTest(FillDatabaseData):
...
+        node.wait_db_up(verbose=True)
+        result = node.remoter.run('rpm -qa scylla\*server')
+        new_ver = result.stdout
+        assert self.orig_ver != self.new_ver, "scylla-server version isn't changed"
+        self.new_ver = new_ver
+
+    def rollback_node(self, node):
+        self.log.info('Rollbacking a Node')
+        upgrade_mode = self.db_cluster.params.get('upgrade_mode', default='smooth')
+        new_introduced_pkgs = self.db_cluster.params.get('new_introduced_pkgs', default='scylla\*-tools-core')
```
Comment (roydahan) [inline upgrade_test.py]: I don't think we should have a default..

File: upgrade_test.py [hunk 4]
```diff
@@ -31,50 +31,105 @@ class UpgradeTest(FillDatabaseData):
...
+        node.remoter.run('sudo chmod 644 /etc/yum.repos.d/scylla.repo')
+        node.remoter.run('sudo yum clean all')
+        if upgrade_mode == 'direct':
+            node.remoter.run('sudo yum remove scylla\* -y')
+            node.remoter.run('sudo yum install {} -y' % node.scylla_pkg())
+        elif upgrade_mode == 'maintenance':
+            node.remoter.run('sudo yum downgrade scylla\*%s -y' % self.orig_ver.split('-')[0])
+        else:
+            if new_introduced_pkgs:
+                node.remoter.run('sudo yum remove %s -y' % new_introduced_pkgs)
```
Comment (roydahan) [inline upgrade_test.py]: Can't we just remove everything that is coming from scylla?
Without this being a parameter...

File: upgrade_test.py [hunk 5]
```diff
@@ -31,50 +31,105 @@ class UpgradeTest(FillDatabaseData):
...
+            node.remoter.run('sudo yum downgrade scylla\* -y')
+            if new_introduced_pkgs:
```
[the remark this answers is not in the document]
Comment (roydahan) [reply upgrade_test.py]: I didn't finish review, stopped here.
Please make the necessary fixes.

**Comment by roydahan** [comment]
@amoskong, please fix according to my comments.
Basically, I want the test to match the following jenkins job and use external parameters as less as we can:
http://jenkins.cloudius-systems.com:8080/job/scylla-staging-rolling-upgrade-gce-clone-roy/

File: upgrade_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (amoskong) [reply upgrade_test.py]: Yes, I already have a solution:
Compare the scylla package list before and after switching the repo file, then we know which package is new introduced.
Then we don't need to manually assign it/them.

File: upgrade_test.py (hunk 3 above)
Comment (amoskong) [reply upgrade_test.py]: Agree

**Comment by amoskong** [comment]
I had reserved two parameters (upgrade_rollback_mode, new_introduced_pkgs), we can auto select it in next PR.

screen shot 2018-05-24 at 1 28 03 am

**Comment by roydahan** [comment]
@amoskong, there is much more, you didn't address most of my comments.
Although they show as outdated... please check it again.

**Comment by amoskong** [comment]
@roydahan I already used this PR to test upgrade of 2.2 / 2018.1, the test scenario works well.
We can consider to merge it.

- http://jenkins.cloudius-systems.com:8080/job/scylla-enterprise-2018.1-rolling-upgrade-gce/
- http://jenkins.cloudius-systems.com:8080/job/scylla-enterprise-2.2-rolling-upgrade-gce/

**Comment by avikivity** [comment]
Let's test upgrades with TLS (both client/server and server/server).

**Comment by amoskong** [comment]
SCT supports the TLS (server-to-server),  client-to-server TLS still has
some problem, we will fix it.

**Comment by amoskong** [comment]
I will recheck again, and reply you.

**Comment by amoskong** [comment]
This PR only changed upgrade_test.py, and add tests/rolling-upgrade-gce.yaml, it won't effect other tests. I'm working on other tasks (ubuntu support/ ubuntu upgrade), it depends on this PR. I want to merge the PR to avoid conflicts.

**Comment by amoskong** [comment]
I won't forget the comments, will review them again and fix by another PR.

**Pull request merged**